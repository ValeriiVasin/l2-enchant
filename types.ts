import { Item } from './items';

export interface Enchant {
  item: Item;
  successRate: number;
  required: Item | Map<Item, number>;
  success: Item;
  fail: Item;
}

export type EnchantMap = Map<Item, Enchant>;

export enum EnchantStrategy {
  // enchant level is decreased by 1 in case of fail
  Decrease = 'Decrease',

  // item is destroyed in case of fail
  Destroy = 'Destroy',

  // enchant level drops to 0 in case of fail
  Drop = 'Drop',

  // one of two items disappears in case of failure
  TwoToOne = 'TwoToOne',

  // enchant, if not - item stays
  Safe = 'Safe',
}
