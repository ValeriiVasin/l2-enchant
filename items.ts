export enum Item {
  Nothing = 'Nothing',
  Adena = 'Адена',

  CloakEnchant = 'Свиток: Модифицировать Плащ',
  Cloak = 'Плащ',
  Cloak_1 = 'Плащ +1',
  Cloak_2 = 'Плащ +2',
  Cloak_3 = 'Плащ +3',
  Cloak_4 = 'Плащ +4',
  Cloak_5 = 'Плащ +5',
  Cloak_6 = 'Плащ +6',
  Cloak_7 = 'Плащ +7',
  Cloak_8 = 'Плащ +8',
  Cloak_9 = 'Плащ +9',
  Cloak_10 = 'Плащ +10',

  TalismanOfAden = 'Талисман Адена',
  TalismanOfAden_1 = 'Талисман Адена +1',
  TalismanOfAden_2 = 'Талисман Адена +2',
  TalismanOfAden_3 = 'Талисман Адена +3',
  TalismanOfAden_4 = 'Талисман Адена +4',
  TalismanOfAden_5 = 'Талисман Адена +5',
  TalismanOfAden_6 = 'Талисман Адена +6',
  TalismanOfAden_7 = 'Талисман Адена +7',
  TalismanOfAden_8 = 'Талисман Адена +8',
  TalismanOfAden_9 = 'Талисман Адена +9',
  TalismanOfAden_10 = 'Талисман Адена +10',

  TalismanOfEva = 'Талисман Евы',
  TalismanOfEva_1 = 'Талисман Евы +1',
  TalismanOfEva_2 = 'Талисман Евы +2',
  TalismanOfEva_3 = 'Талисман Евы +3',
  TalismanOfEva_4 = 'Талисман Евы +4',
  TalismanOfEva_5 = 'Талисман Евы +5',
  TalismanOfEva_6 = 'Талисман Евы +6',
  TalismanOfEva_7 = 'Талисман Евы +7',
  TalismanOfEva_8 = 'Талисман Евы +8',
  TalismanOfEva_9 = 'Талисман Евы +9',
  TalismanOfEva_10 = 'Талисман Евы +10',

  HeroCrown = 'Венец Героя',
  HeroCrown_1 = 'Венец Героя +1',
  HeroCrown_2 = 'Венец Героя +2',
  HeroCrown_3 = 'Венец Героя +3',
  HeroCrown_4 = 'Венец Героя +4',
  HeroCrown_5 = 'Венец Героя +5',
  HeroCrown_6 = 'Венец Героя +6',
  HeroCrown_7 = 'Венец Героя +7',
  HeroCrown_8 = 'Венец Героя +8',
  HeroCrown_9 = 'Венец Героя +9',
  HeroCrown_10 = 'Венец Героя +10',

  DragonBelt = 'Пояс Дракона',
  DragonBelt_1 = 'Пояс Дракона +1',
  DragonBelt_2 = 'Пояс Дракона +2',
  DragonBelt_3 = 'Пояс Дракона +3',
  DragonBelt_4 = 'Пояс Дракона +4',
  DragonBelt_5 = 'Пояс Дракона +5',
  DragonBelt_6 = 'Пояс Дракона +6',
  DragonBelt_7 = 'Пояс Дракона +7',
  DragonBelt_8 = 'Пояс Дракона +8',
  DragonBelt_9 = 'Пояс Дракона +9',
  DragonBelt_10 = 'Пояс Дракона +10',

  DragonEssence = 'Эссенция Дракона',
  DragonVarnish = 'Лак Дракона',
  DragonPendantOne = 'Подвеска Дракона Ур.1',
  DragonPendantOne_1 = 'Подвеска Дракона Ур.1 +1',
  DragonPendantOne_2 = 'Подвеска Дракона Ур.1 +2',
  DragonPendantOne_3 = 'Подвеска Дракона Ур.1 +3',
  DragonPendantOne_4 = 'Подвеска Дракона Ур.1 +4',
  DragonPendantOne_5 = 'Подвеска Дракона Ур.1 +5',

  DragonPendantTwo = 'Подвеска Дракона Ур.2',
  DragonPendantTwo_1 = 'Подвеска Дракона Ур.2 +1',
  DragonPendantTwo_2 = 'Подвеска Дракона Ур.2 +2',
  DragonPendantTwo_3 = 'Подвеска Дракона Ур.2 +3',
  DragonPendantTwo_4 = 'Подвеска Дракона Ур.2 +4',
  DragonPendantTwo_5 = 'Подвеска Дракона Ур.2 +5',

  DragonPendantThree = 'Подвеска Дракона Ур.3',
  DragonPendantThree_1 = 'Подвеска Дракона Ур.3 +1',
  DragonPendantThree_2 = 'Подвеска Дракона Ур.3 +2',
  DragonPendantThree_3 = 'Подвеска Дракона Ур.3 +3',
  DragonPendantThree_4 = 'Подвеска Дракона Ур.3 +4',
  DragonPendantThree_5 = 'Подвеска Дракона Ур.3 +5',

  DragonPendantFour = 'Подвеска Дракона Ур.4',
  DragonPendantFour_1 = 'Подвеска Дракона Ур.4 +1',
  DragonPendantFour_2 = 'Подвеска Дракона Ур.4 +2',
  DragonPendantFour_3 = 'Подвеска Дракона Ур.4 +3',
  DragonPendantFour_4 = 'Подвеска Дракона Ур.4 +4',
  DragonPendantFour_5 = 'Подвеска Дракона Ур.4 +5',

  TalismanOfAuthorityFragment = 'Осколок Талисмана Властителя',
  TalismanOfAuthority_1 = 'Талисман Властителя Ур.1',
  TalismanOfAuthority_2 = 'Талисман Властителя Ур.2',
  TalismanOfAuthority_3 = 'Талисман Властителя Ур.3',
  TalismanOfAuthority_4 = 'Талисман Властителя Ур.4',
  TalismanOfAuthority_5 = 'Талисман Властителя Ур.5',

  NobleTalismanOfAuthority_1 = 'Талисман Благородного Властителя Ур.1',
  NobleTalismanOfAuthority_2 = 'Талисман Благородного Властителя Ур.2',
  NobleTalismanOfAuthority_3 = 'Талисман Благородного Властителя Ур.3',
  NobleTalismanOfAuthority_4 = 'Талисман Благородного Властителя Ур.4',
  NobleTalismanOfAuthority_5 = 'Талисман Благородного Властителя Ур.5',

  RadiantTalismanOfAuthority_1 = 'Искрящийся Талисман Властителя Ур.1',
  RadiantTalismanOfAuthority_2 = 'Искрящийся Талисман Властителя Ур.2',
  RadiantTalismanOfAuthority_3 = 'Искрящийся Талисман Властителя Ур.3',
  RadiantTalismanOfAuthority_4 = 'Искрящийся Талисман Властителя Ур.4',
  RadiantTalismanOfAuthority_5 = 'Искрящийся Талисман Властителя Ур.5',

  Rune_1 = 'Руна Ур.1',
  Rune_2 = 'Руна Ур.2',
  Rune_3 = 'Руна Ур.3',
  Rune_4 = 'Руна Ур.4',
  Rune_5 = 'Руна Ур.5',
  Rune_6 = 'Руна Ур.6',
  Rune_7 = 'Руна Ур.7',
  Rune_8 = 'Руна Ур.8',
  Rune_9 = 'Руна Ур.9',
  Rune_10 = 'Руна Ур.10',
  Rune_11 = 'Руна Ур.11',
  Rune_12 = 'Руна Ур.12',
  Rune_13 = 'Руна Ур.13',

  AgathionCoupon = 'Билет Обмена Агатиона',
  AgathionEnchantScroll = 'Свиток: Модифицировать Агатиона',
  Agathion = 'Агатион',
  Agathion_1 = 'Агатион +1',
  Agathion_2 = 'Агатион +2',
  Agathion_3 = 'Агатион +3',
  Agathion_4 = 'Агатион +4',
  Agathion_5 = 'Агатион +5',
  Agathion_6 = 'Агатион +6',
  Agathion_7 = 'Агатион +7',
  Agathion_8 = 'Агатион +8',
  Agathion_9 = 'Агатион +9',
  Agathion_10 = 'Агатион +10',

  EvaScrollEnchantHairAccessory = 'Свиток Евы: Модифицировать Головной Убор',
  Hat = 'Головной Убор',
  Hat_1 = 'Головной Убор +1',
  Hat_2 = 'Головной Убор +2',
  Hat_3 = 'Головной Убор +3',
  Hat_4 = 'Головной Убор +4',
  Hat_5 = 'Головной Убор +5',
  Hat_6 = 'Головной Убор +6',
  Hat_7 = 'Головной Убор +7',
  Hat_8 = 'Головной Убор +8',
  Hat_9 = 'Головной Убор +9',
  Hat_10 = 'Головной Убор +10',

  // стихийное ожерелье
  BlessedSpiritsVarnish = 'Благословенный Лак Духов',
  IgnisNecklace = 'Ожерелье Игниса',
  IgnisNecklace_1 = 'Ожерелье Игниса +1',
  IgnisNecklace_2 = 'Ожерелье Игниса +2',
  IgnisNecklace_3 = 'Ожерелье Игниса +3',
  IgnisNecklace_4 = 'Ожерелье Игниса +4',
  IgnisNecklace_5 = 'Ожерелье Игниса +5',
  IgnisNecklace_6 = 'Ожерелье Игниса +6',
  IgnisNecklace_7 = 'Ожерелье Игниса +7',
  IgnisNecklace_8 = 'Ожерелье Игниса +8',
  IgnisNecklace_9 = 'Ожерелье Игниса +9',
  IgnisNecklace_10 = 'Ожерелье Игниса +10',

  EWB = 'Свиток: Модифицировать Оружие (Ранг B)',
  BOP = 'Лук Угрозы',
  BOP_1 = 'Лук Угрозы +1',
  BOP_2 = 'Лук Угрозы +2',
  BOP_3 = 'Лук Угрозы +3',
  BOP_4 = 'Лук Угрозы +4',
  BOP_5 = 'Лук Угрозы +5',
  BOP_6 = 'Лук Угрозы +6',
  BOP_7 = 'Лук Угрозы +7',
  BOP_8 = 'Лук Угрозы +8',
  BOP_9 = 'Лук Угрозы +9',
  BOP_10 = 'Лук Угрозы +10',
  BOP_11 = 'Лук Угрозы +11',
  BOP_12 = 'Лук Угрозы +12',
  BOP_13 = 'Лук Угрозы +13',
  BOP_14 = 'Лук Угрозы +14',
  BOP_15 = 'Лук Угрозы +15',
  BOP_16 = 'Лук Угрозы +16',
}
