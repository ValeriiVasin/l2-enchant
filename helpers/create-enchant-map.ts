import { Item } from '../items';
import { EnchantMap, EnchantStrategy, EnchantStrategyConfig } from '../types';

export function createEnchantMap({
  items,
  chances,
  enchantItem = Item.Nothing,
  strategy,
}: EnchantStrategyConfig): EnchantMap {
  const result: EnchantMap = new Map();

  if (items.length !== chances.length + 1) {
    throw new Error('Items should always contain one item more than chances');
  }

  for (let i = 0; i < chances.length; i++) {
    const item = items[i];
    const successItem = items[i + 1];
    const failItem = pickItemOnFail(items, i, strategy);
    const chance = chances[i];

    result.set(item, {
      item,
      required: enchantItem,
      success: successItem,
      fail: failItem,
      successRate: chance,
    });
  }

  return result;
}

function pickItemOnFail(
  items: Item[],
  index: number,
  strategy: EnchantStrategy
): Item {
  switch (strategy) {
    case EnchantStrategy.Destroy:
      return Item.Nothing;
    case EnchantStrategy.Drop:
      return items[0];
    case EnchantStrategy.Decrease:
      return items[Math.max(index - 1, 0)];
  }
}
