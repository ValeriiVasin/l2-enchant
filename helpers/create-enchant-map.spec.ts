import { describe, it } from '@jest/globals';
import { Item } from '../items';
import { EnchantStrategy } from '../types';
import { createEnchantMap } from './create-enchant-map';

describe('create enchant map', () => {
  it('throws when items mismatch with chances', () => {
    const wrapper = () =>
      createEnchantMap({
        strategy: EnchantStrategy.Destroy,
        chances: [1, 2, 3],
        items: [Item.Agathion, Item.Agathion_1],
      });

    expect(wrapper).toThrowError('Items should always contain one item more than chances');
  });

  it('Destroy strategy', () => {
    expect(
      createEnchantMap({
        strategy: EnchantStrategy.Destroy,
        enchantItem: Item.EWB,
        chances: [100, 100, 100, 66],
        items: [Item.BOP, Item.BOP_1, Item.BOP_2, Item.BOP_3, Item.BOP_4],
      }),
    ).toMatchSnapshot();
  });

  it('Decrease strategy', () => {
    expect(
      createEnchantMap({
        strategy: EnchantStrategy.Decrease,
        enchantItem: Item.ScrollEnchantCloak,
        chances: [70, 65, 58],
        items: [Item.Cloak, Item.Cloak_1, Item.Cloak_2, Item.Cloak_3],
      }),
    ).toMatchSnapshot();
  });

  describe('Drop strategy', () => {
    it('drops to first element', () => {
      expect(
        createEnchantMap({
          strategy: EnchantStrategy.Drop,
          enchantItem: Item.BenirEnchantItem,
          items: [Item.Benir_1, Item.Benir_2, Item.Benir_3, Item.Benir_4, Item.Benir_5, Item.Benir_6],
          chances: [50, 45, 40, 35, 50],
        }),
      ).toMatchSnapshot();
    });

    it('drops to boundaries', () => {
      expect(
        createEnchantMap({
          strategy: EnchantStrategy.Drop,
          enchantItem: Item.BenirEnchantItem,
          items: [Item.Benir_1, Item.Benir_2, Item.Benir_3, Item.Benir_4, Item.Benir_5, Item.Benir_6],
          chances: [50, 45, 40, 35, 50],
          boundaries: new Set([Item.Benir_3, Item.Benir_5]),
        }),
      ).toMatchSnapshot();
    });
  });

  it('TwoToOne strategy', () => {
    expect(
      createEnchantMap({
        strategy: EnchantStrategy.TwoToOne,
        items: [Item.Brooch_1, Item.Brooch_2, Item.Brooch_3, Item.Brooch_4, Item.Brooch_5],
        chances: [25, 25, 25, 25],
      }),
    ).toMatchSnapshot();
  });

  it('Safe strategy', () => {
    expect(
      createEnchantMap({
        strategy: EnchantStrategy.Safe,
        enchantItem: Item.StableScrollEnchantLegendaryCloak,
        chances: [28, 28, 48, 48, 20],
        items: [
          Item.LegendaryCloak_10,
          Item.LegendaryCloak_11,
          Item.LegendaryCloak_12,
          Item.LegendaryCloak_13,
          Item.LegendaryCloak_14,
          Item.LegendaryCloak_15,
        ],
      }),
    ).toMatchSnapshot();
  });
});
