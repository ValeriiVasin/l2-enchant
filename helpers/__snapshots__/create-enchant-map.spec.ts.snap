// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create enchant map Decrease strategy 1`] = `
Map {
  "Плащ" => Object {
    "cost": "Свиток: Модифицировать Плащ",
    "item": "Плащ",
    "results": Array [
      Object {
        "chance": 70,
        "item": "Плащ +1",
      },
      Object {
        "chance": 30,
        "item": "Плащ",
      },
    ],
  },
  "Плащ +1" => Object {
    "cost": "Свиток: Модифицировать Плащ",
    "item": "Плащ +1",
    "results": Array [
      Object {
        "chance": 65,
        "item": "Плащ +2",
      },
      Object {
        "chance": 35,
        "item": "Плащ",
      },
    ],
  },
  "Плащ +2" => Object {
    "cost": "Свиток: Модифицировать Плащ",
    "item": "Плащ +2",
    "results": Array [
      Object {
        "chance": 58,
        "item": "Плащ +3",
      },
      Object {
        "chance": 42,
        "item": "Плащ +1",
      },
    ],
  },
}
`;

exports[`create enchant map Destroy strategy 1`] = `
Map {
  "Лук Угрозы" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы",
    "results": Array [
      Object {
        "chance": 100,
        "item": "Лук Угрозы +1",
      },
      Object {
        "chance": 0,
        "item": "Nothing",
      },
    ],
  },
  "Лук Угрозы +1" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы +1",
    "results": Array [
      Object {
        "chance": 100,
        "item": "Лук Угрозы +2",
      },
      Object {
        "chance": 0,
        "item": "Nothing",
      },
    ],
  },
  "Лук Угрозы +2" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы +2",
    "results": Array [
      Object {
        "chance": 100,
        "item": "Лук Угрозы +3",
      },
      Object {
        "chance": 0,
        "item": "Nothing",
      },
    ],
  },
  "Лук Угрозы +3" => Object {
    "cost": "Свиток: Модифицировать Оружие (Ранг B)",
    "item": "Лук Угрозы +3",
    "results": Array [
      Object {
        "chance": 66,
        "item": "Лук Угрозы +4",
      },
      Object {
        "chance": 34,
        "item": "Nothing",
      },
    ],
  },
}
`;

exports[`create enchant map Drop strategy drops to boundaries 1`] = `
Map {
  "Талисман Бенира Ур.1" => Object {
    "cost": "Осколок Талисмана Бенира",
    "item": "Талисман Бенира Ур.1",
    "results": Array [
      Object {
        "chance": 50,
        "item": "Талисман Бенира Ур.2",
      },
      Object {
        "chance": 50,
        "item": "Талисман Бенира Ур.1",
      },
    ],
  },
  "Талисман Бенира Ур.2" => Object {
    "cost": "Осколок Талисмана Бенира",
    "item": "Талисман Бенира Ур.2",
    "results": Array [
      Object {
        "chance": 45,
        "item": "Талисман Бенира Ур.3",
      },
      Object {
        "chance": 55,
        "item": "Талисман Бенира Ур.1",
      },
    ],
  },
  "Талисман Бенира Ур.3" => Object {
    "cost": "Осколок Талисмана Бенира",
    "item": "Талисман Бенира Ур.3",
    "results": Array [
      Object {
        "chance": 40,
        "item": "Талисман Бенира Ур.4",
      },
      Object {
        "chance": 60,
        "item": "Талисман Бенира Ур.3",
      },
    ],
  },
  "Талисман Бенира Ур.4" => Object {
    "cost": "Осколок Талисмана Бенира",
    "item": "Талисман Бенира Ур.4",
    "results": Array [
      Object {
        "chance": 35,
        "item": "Талисман Бенира Ур.5",
      },
      Object {
        "chance": 65,
        "item": "Талисман Бенира Ур.3",
      },
    ],
  },
  "Талисман Бенира Ур.5" => Object {
    "cost": "Осколок Талисмана Бенира",
    "item": "Талисман Бенира Ур.5",
    "results": Array [
      Object {
        "chance": 50,
        "item": "Талисман Бенира Ур.6",
      },
      Object {
        "chance": 50,
        "item": "Талисман Бенира Ур.5",
      },
    ],
  },
}
`;

exports[`create enchant map Drop strategy drops to first element 1`] = `
Map {
  "Талисман Бенира Ур.1" => Object {
    "cost": "Осколок Талисмана Бенира",
    "item": "Талисман Бенира Ур.1",
    "results": Array [
      Object {
        "chance": 50,
        "item": "Талисман Бенира Ур.2",
      },
      Object {
        "chance": 50,
        "item": "Талисман Бенира Ур.1",
      },
    ],
  },
  "Талисман Бенира Ур.2" => Object {
    "cost": "Осколок Талисмана Бенира",
    "item": "Талисман Бенира Ур.2",
    "results": Array [
      Object {
        "chance": 45,
        "item": "Талисман Бенира Ур.3",
      },
      Object {
        "chance": 55,
        "item": "Талисман Бенира Ур.1",
      },
    ],
  },
  "Талисман Бенира Ур.3" => Object {
    "cost": "Осколок Талисмана Бенира",
    "item": "Талисман Бенира Ур.3",
    "results": Array [
      Object {
        "chance": 40,
        "item": "Талисман Бенира Ур.4",
      },
      Object {
        "chance": 60,
        "item": "Талисман Бенира Ур.1",
      },
    ],
  },
  "Талисман Бенира Ур.4" => Object {
    "cost": "Осколок Талисмана Бенира",
    "item": "Талисман Бенира Ур.4",
    "results": Array [
      Object {
        "chance": 35,
        "item": "Талисман Бенира Ур.5",
      },
      Object {
        "chance": 65,
        "item": "Талисман Бенира Ур.1",
      },
    ],
  },
  "Талисман Бенира Ур.5" => Object {
    "cost": "Осколок Талисмана Бенира",
    "item": "Талисман Бенира Ур.5",
    "results": Array [
      Object {
        "chance": 50,
        "item": "Талисман Бенира Ур.6",
      },
      Object {
        "chance": 50,
        "item": "Талисман Бенира Ур.1",
      },
    ],
  },
}
`;

exports[`create enchant map Safe strategy 1`] = `
Map {
  "Легендарный Плащ +10" => Object {
    "cost": "Стабильный Свиток: Модифицировать Легендарный Плащ",
    "item": "Легендарный Плащ +10",
    "results": Array [
      Object {
        "chance": 28,
        "item": "Легендарный Плащ +11",
      },
      Object {
        "chance": 72,
        "item": "Легендарный Плащ +10",
      },
    ],
  },
  "Легендарный Плащ +11" => Object {
    "cost": "Стабильный Свиток: Модифицировать Легендарный Плащ",
    "item": "Легендарный Плащ +11",
    "results": Array [
      Object {
        "chance": 28,
        "item": "Легендарный Плащ +12",
      },
      Object {
        "chance": 72,
        "item": "Легендарный Плащ +11",
      },
    ],
  },
  "Легендарный Плащ +12" => Object {
    "cost": "Стабильный Свиток: Модифицировать Легендарный Плащ",
    "item": "Легендарный Плащ +12",
    "results": Array [
      Object {
        "chance": 48,
        "item": "Легендарный Плащ +13",
      },
      Object {
        "chance": 52,
        "item": "Легендарный Плащ +12",
      },
    ],
  },
  "Легендарный Плащ +13" => Object {
    "cost": "Стабильный Свиток: Модифицировать Легендарный Плащ",
    "item": "Легендарный Плащ +13",
    "results": Array [
      Object {
        "chance": 48,
        "item": "Легендарный Плащ +14",
      },
      Object {
        "chance": 52,
        "item": "Легендарный Плащ +13",
      },
    ],
  },
  "Легендарный Плащ +14" => Object {
    "cost": "Стабильный Свиток: Модифицировать Легендарный Плащ",
    "item": "Легендарный Плащ +14",
    "results": Array [
      Object {
        "chance": 20,
        "item": "Легендарный Плащ +15",
      },
      Object {
        "chance": 80,
        "item": "Легендарный Плащ +14",
      },
    ],
  },
}
`;

exports[`create enchant map TwoToOne strategy 1`] = `
Map {
  "Брошь Ур.1" => Object {
    "cost": "Брошь Ур.1",
    "item": "Брошь Ур.1",
    "results": Array [
      Object {
        "chance": 25,
        "item": "Брошь Ур.2",
      },
      Object {
        "chance": 75,
        "item": "Брошь Ур.1",
      },
    ],
  },
  "Брошь Ур.2" => Object {
    "cost": "Брошь Ур.2",
    "item": "Брошь Ур.2",
    "results": Array [
      Object {
        "chance": 25,
        "item": "Брошь Ур.3",
      },
      Object {
        "chance": 75,
        "item": "Брошь Ур.2",
      },
    ],
  },
  "Брошь Ур.3" => Object {
    "cost": "Брошь Ур.3",
    "item": "Брошь Ур.3",
    "results": Array [
      Object {
        "chance": 25,
        "item": "Брошь Ур.4",
      },
      Object {
        "chance": 75,
        "item": "Брошь Ур.3",
      },
    ],
  },
  "Брошь Ур.4" => Object {
    "cost": "Брошь Ур.4",
    "item": "Брошь Ур.4",
    "results": Array [
      Object {
        "chance": 25,
        "item": "Брошь Ур.5",
      },
      Object {
        "chance": 75,
        "item": "Брошь Ур.4",
      },
    ],
  },
}
`;
